### ISSUE TITLE
Support using `Base` for initialization in `init` function

### ISSUE THREAD

**Issue by lilizoey** [godot-rust/gdext#557]
When initializing an object in Godot, it is a fairly common pattern in c++ to do most of the setup in the `init` function. But as of #497 and #501, it is now impossible to directly use the `Base` object to do anything (at least without using `#[doc(hidden)]` functions). That means that in order to call any methods on base to set up the class, you'd need to first create the `Self` object and then use a method like `WithBaseField::to_gd`, like this:

```rs
fn init(base: Base<Self::Base>) -> Self {
  let self_ = Self { .., base };
  self_.to_gd().some_method();
}
```

However this means that every field in `Self` has some sensible default, that doesn't require any calls to the base object. It would be nice to provide an official solution for this pattern.

This was initially discussed in this discord thread https://discord.com/channels/723850269347283004/1193536016968073236, i will summarize the solutions we discussed here.

## Possible Solutions
1. Make `Base`'s `to_gd` method public. This method is currently `#[doc(hidden)]`, but using it solves the issue. However the reason it is hidden is because it may cause confusion in most other functions, where it may be unclear if people should use `self.base.to_gd()` or `self.to_gd()` or even `self.base_mut()`. As well as what the difference is. So if we do this we should likely rename it to make it more clear.
2. Have a "transport base" and a "stored base". Here we would add a new type (the transport base) that can be used to call methods on base, and it is what is passed to `init`. Whereas the stored base is what we currently have and is what is stored in the struct. This would however require us adding a new type that is only used in this one situation.
3. Pass in a `Gd` of base to the init function. This might however be confusing as both the `Base` and `Gd` point to the same thing, and people might be tempted to store the wrong object in the struct. 
4. Add a post-init function (possibly replacing the original init function). This does make it easy to do initialization using the base object, however it doesn't solve the issue that every field needs a sensible default.

I am currently inclined to go with the first option, to make `to_gd` public. It seems the simplest to me and giving it a new name with good documentation would hopefully solve the confusion issue.

**Comment by Bromeon** [comment]
Another potential option:

5. Have `Base` in two possible states: transport + initialized.
   * In the `transport` state, one can call methods on it directly. We could allow `to_gd()` only in that phase, maybe with a better name, such as `during_init()` or whatever.
   * In the `initialized` phase, it would behave exactly like today, but `during_init()` would panic with an expressive message.

In other words, type-state without introducing yet another type (which is also an option, but really makes the API even more complex).

**Comment by MatrixDev** [comment]
I'll just copy it here from my duplicate issue so it would not be lost.

My case was a little different, I want to have a fully constructed tree when `init` finishes but it has a lot in common. One more solution is to add new or change `init` function:

```rust
fn init() -> Gd<Self> {
    let mut node = Node3D::new_alloc();
    // do something with node

    let mut this = Gd::from_init_fn(|base|
        Self {
            node: node.clone(),
            base,
        }
    );
    this.base_mut().add_child(node.upcast());
    this
}
```

**Comment by mhgolkar** [comment]
> My case was a little different, ...

Hi @MatrixDev 

Why not implementing it this way without `from_init_fn`:

```Rust
fn init(base: Base</*Super*/Node>) -> Self /* Extended */ {
    let mut base_gd = base.to_gd();
    let inner_node = Node3D::new_alloc();
    // ... manipulate the inner node:
    // e.g. inner_node.set("foo_property", bar_variant);
    // ...
    base_gd.add_child(inner_node.clone().upcast());
    // ...
    Self {
        base,
        inner_node
    }
}
```

Just a side note: Currently `.to_gd` is `#[doc(hidden)]` so your IDE may not suggest it, but it's going to be there!

**Comment by Bromeon** [comment]
> but it's going to be there!

There's no guarantee. `#[doc(hidden)]` is explicitly [**NOT** part of the public API](https://godot-rust.github.io/docs/gdext/master/godot/#public-api). 
So we do not take special precautions when it comes to breaking code.

**Comment by mhgolkar** [comment]
> There's no guarantee. `#[doc(hidden)]` is explicitly [**NOT** part of the public API](https://godot-rust.github.io/docs/gdext/master/godot/#public-api).

Sure! I should have added "for the time being"!

**Comment by MatrixDev** [comment]
> Sure! I should have added "for the time being"!

I'm currently doing exactly this. But my issue was `No recommended way to call Base<T> methods from init`.

**Comment by RedMser** [comment]
In case anyone stumbles upon this issue, it's a bit different for RefCounted classes like Resources.
I only got it to work by intentionally leaking memory, otherwise the parent reference always ended up as a nullptr 🔥

```rs
#[derive(GodotClass)]
#[class(base=Translation)]
pub struct TranslationFluent {
    base: Base<Translation>,
}

#[godot_api]
impl ITranslation for TranslationFluent {
    fn init(base: Base<Translation>) -> Self {
        std::mem::forget(base.to_gd()); // <-- this was needed to fix crashes
        base.to_gd().set_locale("".into());

        Self {
            base,
        }
    }
}
```

**Comment by fpdotmonkey** [comment]
Does the gdextension team have a position on what should go in `init` vs `ready`?  It could be that the gdextension API was intended to be like gdscript in the sense of most godot initialization operations should happen on `ready`.  The argument in favor of doing more initialization on `init` is that that's how engine classes do it, but they critically are not built with gdextension.

**Comment by RedMser** [comment]
> to be like gdscript in the sense of most godot initialization operations should happen on ready. 

That's simply not true. A primary use case is any class that does not inherit Node, does not have a ready function. Initializing it must be done within a custom constructor or init.

See my code snippet, I'd like for all created Translations to start with an empty locale, overriding the default of "en". There is no other way to do this in a way that Godot knows that this is the new default value (and won't show a revert arrow in the inspector).

**Comment by lilizoey** [comment]
I think the issue with `RefCounted` classes can be fixed relatively easily by just initializing the refcount earlier in the construction process. Since currently the refcount isn't initialized by the time `init` is called.

**Comment by lilizoey** [comment]
So that does sorta fix the issue, but there are some memory leaks and i dont think there is a way to fix those. Since if the refcount is initialized by `init` then if a class is created from godot then we just cant tell that the refcount shouldn't be incremented.

The fundamental problem with this is that our `Gd` will try to do reference counting, but specifically in the `init` function we must avoid reference counting the base object. But as soon as we let people access the base object as a `Gd` it is impossible to prevent them from just getting a `Gd` out of it which will attempt to reference count stuff.

**Comment by lilizoey** [comment]
I guess one thing that could be doable is, initialize the refcount early. And then keep track in `Base` if the refcount needs to be decremented at some point, and find an appropriate place which we know for sure will be called _after_ initialization at some point for all classes in which we can decrement the refcount.

**Comment by goatfryed** [comment]
Regarding option 5 with a base that's initialization state aware

What do you think about exposing this via `.base()`  so the following works

```rs
let mut this = Self {
     rd: Option::None,
     shader: Rid::Invalid,
     pipeline: Rid::Invalid,
     base,
 };

 this.base_mut()
     .set_effect_callback_type(EffectCallbackType::POST_TRANSPARENT);

 this
```

At the moment this fails due to some downcast panic, so maybe it could switch their depending on the state.

```
  downcast from Node to NeutralLocationState failed; instance Gd { id: 30249321849, class: Node }
  Context: panic during NeutralLocationState::init() constructor
```

This might avoid the introduction of a new type, not pollute the public api outside of init, is something that feels intuitive (at least to me)  and might be also easy to implement compared to new types and so on. It also scopes the problematic api to the init, where I would expect some partial api behavior.

I'm also thinking about potential confusing cases like

```rs
fn init(base: Base</*Super*/Node>) -> Self /* Extended */ {
    let my_node = MyNode::new_alloc();
    my_node.during_init().set_value("foobar");

    Self {
        base,
        my_node,
    }
}
```

where someone uses the wrong api because it's during init, but not that node's init.

**Comment by Bromeon** [comment]
> What do you think about exposing this via `.base()` so the following works
> 
> ```rs
> let mut this = Self {
>      rd: Option::None,
>      shader: Rid::Invalid,
>      pipeline: Rid::Invalid,
>      base,
>  };
> 
>  this.base_mut()
>      .set_effect_callback_type(EffectCallbackType::POST_TRANSPARENT);
> 
>  this
> ```

This might be one way, but why not 

```rs
base.during_init()
    .set_effect_callback_type(EffectCallbackType::POST_TRANSPARENT);

Self { /* other args */, base }
```

I'm not fully convinced that we should support operations like `WithBaseField::base_mut()` on half-initialized objects... It can create rather confusing scenarios because it looks like regular code during the lifetime of a `Gd<T>`, but many operations are in fact not available. It's probably cleaner to treat the base as an explicit `Gd<Base>` object -- which it _is_ at that point in time.

For example, there are operations you're not allowed to do in a C++ constructor because of exactly this issue (base object is ready, but current object is still being constructed), for example virtual method calls. Pretending that an object were initialized while it's not might bring these issues to Rust.

---



> I'm also thinking about potential confusing cases like [...] where someone uses the wrong api because it's during init, but not that node's init.

`during_init()` would exist only on `Base<T>`, not `Gd<T>`, so this can't happen. But apart from that, I'm not sure if "someone might call method on the wrong object" is a very strong argument.

**Comment by TitanNano** [comment]
This probably an unpopular opinion, but I think the “right” way in rust would be to pass a `Gd<T>` into the `init` function and then convert it into a `Base<T>` when assigning it to `Self`.

**Comment by Bromeon** [comment]
> This probably an unpopular opinion, but I think the “right” way in rust would be to pass a `Gd<T>` into the `init` function and then convert it into a `Base<T>` when assigning it to `Self`.

Also an option, but that would make the very common case (just forwarding `base`) harder. It would also encourage to treat the base as regular `Gd<T>` (e.g. store it somewhere), which may cause other issues.

A `base.during_init()` is more explicit in that regard. It could even just give access to `&Gd` or `&mut Gd`, or a with-style closure, so people need to go to extraordinary lengths (clone) to store it. But not sure if that's necessary...

**Comment by goatfryed** [comment]
> But apart from that, I'm not sure if "someone might call method on the wrong object" is a very strong argument

Totally true. It is a weak argument. It's a rare use case in a rare use case anyway. It's only a help request concern what implementation might lead to users requesting help, because they don't understand the error in their operation.

On the other hand some changes that introduce more issues with half initialized would be a more relevant concern. I take back my suggestion.

**Comment by TitanNano** [comment]
> It would also encourage to treat the base as regular Gd<T> (e.g. store it somewhere), which may cause other issues.

How is it different from storing an upcasted `Gd<Self>`? 

It might confuse people that don't read the documentation and wonder what exactly to do with the `Gd<T>` since having a `base: Base<T>` is not strictly mandatory.

Thinking about it more, this would make the `Gd` to `Base` conversion public, which should probably not be possible on arbitrary `Gd`s... So the right solution would be to add _yet another_ type like `PendingBase` which can be used like a Gd but is the only thing that can be converted to `Base`. But yeah, that makes things more complicated for the common, straightforward cases. 

> A base.during_init() is more explicit in that regard.

Yes, but what irritates me a bit is that it will introduce a (seemingly unnecessary) runtime state check that might panic.

**Comment by Bromeon** [comment]
> How is it different from storing an upcasted `Gd<Self>`?

You don't have access to `Gd<Self>` during the constructor, not sure what you mean?
Once the object is constructed and thus embedded in `Gd`, it's no longer an issue...

---

> It might confuse people that don't read the documentation and wonder what exactly to do with the `Gd<T>` since having a `base: Base<T>` is not strictly mandatory.

Yes, I think `Base` being a type distinct from `Gd` helps in this regard (that was the original intention).

---

> Thinking about it more, this would make the `Gd` to `Base` conversion public, which should probably not be possible on arbitrary `Gd`s... So the right solution would be to add _yet another_ type like `PendingBase` which can be used like a Gd but is the only thing that can be converted to `Base`. But yeah, that makes things more complicated for the common, straightforward cases.

This is what I mentioned [above](https://github.com/godot-rust/gdext/issues/557#issuecomment-1880109912) with:

> introducing yet another type (which is also an option, but really makes the API even more complex)

It's an edge case, and we already have `Gd`-like type proliferation (`Gd`, `DynGd`, `Base`, soon `UniqueGd` and possibly more for multithreading). It also mandates a conversion that's not necessary in 95% of cases where the base is simply forwarded.

---

> Yes, but what irritates me a bit is that it will introduce a (seemingly unnecessary) runtime state check that might panic.

We can disable the check under the Release profile. And it cannot panic within `init()`.

If someone calls a method that is deliberately named `during_init()` not _during init_, they absolutely deserve the panic 😁

**Comment by Yarwin** [comment]
> What do you think about exposing this via .base() so the following works
> 
> ```rs
> let mut this = Self {
>      rd: Option::None,
>      shader: Rid::Invalid,
>      pipeline: Rid::Invalid,
>      base,
>  };
> 
>  this.base_mut()
>      .set_effect_callback_type(EffectCallbackType::POST_TRANSPARENT);
> 
>  this
> ```

Stupid question, but doesn't `call_deferred` already cover such cases?

**Comment by TitanNano** [comment]
> It's an edge case, and we already have Gd-like type proliferation (Gd, DynGd, Base, soon UniqueGd and possibly more for multithreading).

Yes I was circling back to what you already mentioned. 

> It also mandates a conversion that's not necessary in 95% of cases where the base is simply forwarded.

So, do we actually need this? Personally, I never had an issue with this and always found a way around the limitation. As @Yarwin mentions, there is `call_deferred` and for nodes one can use `ready`. I wonder if the benefit really outweighs the complexity.

**Comment by Bromeon** [comment]
> So, do we actually need this?

Yes. It has been requested many times, here (#585 #997) and on Discord ([1](https://discord.com/channels/723850269347283004/1193536016968073236/1193536016968073236) [2](https://discord.com/channels/723850269347283004/1381967978521956382/1381967978521956382) ...), by at least @bluenote10 @RedMser @malj @MatrixDev @mhgolkar @goatfryed.

Accessing the base object before construction is also safe, what's problematic is accessing it through the not-yet-initialized derived object (i.e. `base()` + `base_mut()`).

**Comment by Bromeon** [comment]
Accessing the `Gd` during init is the easy part, but upon re-reading the discussion, I discovered [this](https://github.com/godot-rust/gdext/issues/557#issuecomment-2083326637):

> In case anyone stumbles upon this issue, it's a bit different for RefCounted classes like Resources.
> I only got it to work by intentionally leaking memory, otherwise the parent reference always ended up as a nullptr 🔥

This is a real problem -- handing out `Gd<RefCounted>` pointers during construction will cause UAF (because `Base` holds only a weak pointer). Incrementing ref-count will cause a leak later. I'm already working on an implementation, but need to address this somehow...

**Comment by Yarwin** [comment]
> (…) there is call_deferred and for nodes one can use ready. I wonder if the benefit really outweighs the complexity.

I finally checked it and using `call_deferred` on non-initialized base will cause the panic; it doesn't apply to Callable obviously. 

> This is a real problem -- handing out Gd<RefCounted> pointers during construction will cause UAF (because Base holds only a weak pointer). Incrementing ref-count will cause a leak later. I'm already working on an implementation, but need to address this somehow...

I was experimenting with callbacks and refcounts are weird – editor might want to create placeholder/default instances and working with them is straight way to UB. (Hell, calling `base()` on constructed instance with RefCounted `Base` in `init` is enough to cause an UB already 😅)

In other words, an example adapted from https://github.com/helf4ch/shader_test: (ofc `to_gd_weak` and `instance_id` are not present in base/current codebase, I created them ad-hoc to poke the editor; anything like `to_gd_weak` should NEVER be exposed as a part of user-facing API because it is unsound)

```rs
    fn init(base: Base<CompositorEffect>) -> Self {

        // Use async to await one frame and then do stuff on our initialized instance.
        unsafe {
            if let Some(tree) = Engine::singleton().get_main_loop().map(Gd::cast::<SceneTree>) {
                let instance_id = base.instance_id();
                godot_print!("Ran init for instance: {instance_id}");
                let weakref = base.to_gd_weak();
                

                godot::task::spawn(async move {
                    let _ = tree.signals().process_frame().to_future().await;
                    // For some unknown reason editor might want to create placeholder instances which will be dropped instantly after init.
                    let Ok(mut instance) = Gd::<GrayscaleEffect>::try_from_instance_id(instance_id) else {
                        godot_print!("Null instance: {instance_id}");
                        return
                    };
                    // Should be two – one for our instance held by editor, and another held by `instance` variable.
                    godot_print!("Refcount: {}", weakref.get_reference_count());
                    weakref.to_variant().call("reference", &[]);
                    instance.set_effect_callback_type(EffectCallbackType::POST_TRANSPARENT);
                    godot_print!("Refcount: {}", weakref.get_reference_count());
                });
            }
        }

        Self {
            rd: Option::None,
            shader: Rid::Invalid,
            pipeline: Rid::Invalid,
            base,
        }
    }

```

works weirdly:

![Image](https://github.com/user-attachments/assets/4f9acae4-b930-4c5e-a937-ba2c5e1c7169)

Firstly Godot creates placeholder/base instance, which is being discarded imminently. Our closure fails and `weakref` is being dropped. It doesn't cause any UB, because `weakref` points to null instance that has been already freed 🙃 (cool).

Afterwards Godot creates "real" instance which is being kept. Everything is good – I increase refcount for our weakref so it is properly decreased upon `Drop`.

What works nicely and reliably is relying on `InstanceId` alone.

```rs

        if let Some(tree) = Engine::singleton().get_main_loop().map(Gd::cast::<SceneTree>) {
            let instance_id = base.instance_id();
            godot_print!("Ran init for instance: {instance_id}");

            godot::task::spawn(async move {
                let _ = tree.signals().process_frame().to_future().await;
                // For some unknown reason editor might want to create placeholder instances which will be dropped instantly after init.
                let Ok(mut instance) = Gd::<GrayscaleEffect>::try_from_instance_id(instance_id) else {
                    godot_print!("Null instance: {instance_id}");
                    return
                };
                instance.set_effect_callback_type(EffectCallbackType::POST_TRANSPARENT);
            });
        }

```

![Image](https://github.com/user-attachments/assets/6cf6f7eb-5044-43c3-a64d-10bc783d1025)

("default" instance is being created only once and kept later.)

IMO easiest way to handle that would be creating an additional custom callback method for ITraits which we would call deferred after `init` – and which would get access to `self` by constructing it from `InstanceId`.

Something along the lines of

```rs
#[godot_api]
impl IResource for MyRes {
fn post_init(&mut self) {...}
}
```

ain't _that_ bad (it is kind of bad, since it is not present in godot at all + overlaps with ready, but gdscript does similar shenanigans with its init/ready so ehhhhh).

**Comment by TitanNano** [comment]
> IMO easiest way to handle that would be creating an additional custom callback method for ITraits which we would call deferred after init – and which would get access to self by constructing it from InstanceId.

There is also the post_init notification from the engine... perhaps we can just hook that up to a callback method.

**Comment by sylbeth** [comment]
So, trying to do some work on init, I stumbled upon the Editor gaslighting me:

```rust
fn init(base: Base<RigidBody3D>) -> Self {
    let mut _base = base.to_gd();
    _base.set_contact_monitor(true);
    _base.set_max_contacts_reported(8);
    let mut sprite = Sprite3D::new_alloc();
    sprite.set_billboard_mode(BillboardMode::ENABLED);
    _base.add_child(&sprite);
    Self {
        base,
        sprite,
        texture: OnEditor::default(),
    }
}
```

This initialization sets a default value for contact_monitor and max_contacts_reported, but the moment you open a scene with them, the editor modifies the scene so that the value of those parameters go back to the defaults. The only way to fix it that I found is either make it a tool script or not opening the scenes where this node exists in.

**Comment by Yarwin** [comment]
> So, trying to do some work on init, I stumbled upon the Editor gaslighting me:
> 
> (...)
> 
> This initialization sets a default value for contact_monitor and max_contacts_reported, but the moment you open a scene with them, the editor modifies the scene so that the value of those parameters go back to the defaults. The only way to fix it that I found is either make it a tool script or not opening the scenes where this node exists in.

It is actually unrelated to this issue – usually editor doesn't hold (and operate on) "real" non-tool instances, it replaces it with placeholders.

for example, in such a setup (non tool Node as a root of the scene in the editor + Callable executed in the editor) the unreachable!() block will never be reached when working with non-tool object (it is not guaranteed though; the inner workings and various side effects are not documented and effects might wary depending on Godot version). 

```rs
    fn init(base: Base<Self::Base>) -> Self {
        let  this = Self {why: 42, base};
        let mut obj = this.base().clone();

        Callable::from_local_fn("do stuff in editor without being a tool", move |_| {
            if !obj.is_instance_valid() {
                godot_print!("object invalid, aborting");
                return Ok(Variant::nil())
            }
            unreachable!();
            obj.set_contact_monitor(true);
            obj.set_max_contacts_reported(8);
            obj.notify_property_list_changed();
            Ok(Variant::nil())
        }).call_deferred(&[]);

        this
    }
```

> the moment you open a scene with them, the editor modifies the scene so that the value of those parameters go back to the defaults. 

They shouldn't set default values 🤔 (and if they do, they should be written in scene file).

I haven't checked if it differs if they are instantiated by other tool node though; it might be different if given node is not a root node of the scene as well.

Additionally `add_child(...)` might cause a memory leak.

In general – in such a case use tools (pun not intended) provided by Godot (editor plugins, instantiating with other node, `tool` etc), hacking in `init` is not a good (nor stable) solution.

**Comment by Bromeon** [comment]
@Yarwin's answer is a very thorough one 👍 From now on, let's please keep the issue focused on the "access base during init" problem. I marked the two as off-topic to keep this thread readable for users visiting for the original problem.

There's also [a parallel Discord thread](https://discord.com/channels/723850269347283004/1396472640599883817) that you opened with the same question -- we can continue there.
